core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -1399,6 +1397,158 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+    ).asJavaCollection
+
+    var alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1,
+      topicResource2 -> topicConfigEntries2
+    ).asJava)
+
+    assertEquals(Set(topicResource1, topicResource2).asJava, alterResult.values.keySet)
+    alterResult.all.get
+
+    // Verify that topics were updated correctly
+    var describeResult = client.describeConfigs(Seq(topicResource1, topicResource2).asJava)
+    var configs = describeResult.all.get
+
+    assertEquals(2, configs.size)
+
+    assertEquals("1000", configs.get(topicResource1).get(LogConfig.FlushMsProp).value)
+    assertEquals("compact,delete", configs.get(topicResource1).get(LogConfig.CleanupPolicyProp).value)
+    assertEquals((Defaults.LogRetentionHours * 60 * 60 * 1000).toString, configs.get(topicResource1).get(LogConfig.RetentionMsProp).value)
+
+    assertEquals("0.9", configs.get(topicResource2).get(LogConfig.MinCleanableDirtyRatioProp).value)
+    assertEquals("lz4", configs.get(topicResource2).get(LogConfig.CompressionTypeProp).value)
+
+    //verify config value substract
+    topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.SUBSTRACT)
+    ).asJava
+
+   alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1
+    ).asJava)
+    alterResult.all.get
+
+    // Verify that topics were updated correctly
+    describeResult = client.describeConfigs(Seq(topicResource1).asJava)
+    configs = describeResult.all.get
+
+    assertEquals("delete", configs.get(topicResource1).get(LogConfig.CleanupPolicyProp).value)
+
+    // Alter topics with validateOnly=true
+    topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.APPEND)
+    ).asJava
+
+    alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1
+    ).asJava, new IncrementalAlterConfigsOptions().validateOnly(true))
+    alterResult.all.get
+
+    // Verify that topics were not updated due to validateOnly = true
+    describeResult = client.describeConfigs(Seq(topicResource1).asJava)
+    configs = describeResult.all.get
+
+    assertEquals("delete", configs.get(topicResource1).get(LogConfig.CleanupPolicyProp).value)
+
+    //Alter topics with validateOnly=true with invalid configs
+    topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CompressionTypeProp, "zip"), AlterConfigOp.OpType.SET)
+    ).asJava
+
+    alterResult = client.incrementalAlterConfigs(Map(
+      topicResource1 -> topicConfigEntries1
+    ).asJava, new IncrementalAlterConfigsOptions().validateOnly(true))
+
+    assertFutureExceptionTypeEquals(alterResult.values().get(topicResource1), classOf[InvalidRequestException],
+      Some("Invalid config value for resource"))
+  }
```

Comment:
Should we also add a test case against lost updates? e.g `a=1,b=2,c=3`, incremental config change - `a=4,b=5` and we ensure that `c=3` still

Reply:
I think this can piggy-back on the `//verify config value substract` check

Reply:
added a check to verify previous updates are intact.